### PR TITLE
Document the feedback spam classifier

### PR DESCRIPTION
**Pull request by jonathanglassman**
Document the feedback spam classifier

### CODE REVIEW

File: source/analysis/spam-filter/index.html.md.erb [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+- searches the text in the Pandas dataframe for features that may indicate that the feedback data is spam (for example, if the feedback text has SQL)
+
+### Run script 002
+
+Run the [`002_train.py` Python script](https://github.com/alphagov/govuk-feedback-spam-classifier/blob/master/src/002_train.py) in your virtual environment.
+
+This script:
+
+- [trains the spam classifier model](https://en.wikipedia.org/wiki/Training,_validation,_and_test_data_sets) using the specified data set
+- creates visualisations to help you assess if the model is classifying spam correctly
```
Comment (FelixReillyGDS) [inline source/analysis/spam-filter/index.html.md.erb:63]: For this, I think it would be more accurate to say something like "visualisations to demonstrate the accuracy of the model's predictions in comparison to the manual labels from the training dataset, and the features that are most 'important' to the model's classification decisions" - @hwrightson let me know!

File: source/analysis/spam-filter/index.html.md.erb [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+1. Clone the [feedback spam classifier repo](https://github.com/alphagov/govuk-feedback-spam-classifier) to your local machine.
+
+## Run the feedback spam classifier Python scripts
+
+You run 4 Python scripts in your virtual environment to try the feedback spam classifier. These scripts are set up to use the default feedback data set.
+
+If you want to try the feedback spam classifier with a different data set, you must:
+
+- specify where this different data set is in the [params.yaml file in the feedback spam classifier repo](https://github.com/alphagov/govuk-feedback-spam-classifier/blob/master/params.yaml)
+- make sure that this different data set is similar in format and structure to the default data set
```
Comment (FelixReillyGDS) [inline source/analysis/spam-filter/index.html.md.erb:45]: Possibly needs an extra point to specify that commands such as "features_df = features_df.drop(["created_at", "day", "month", "description", "url"], axis=1)"  in "002_train.py" will have to be updated to align with the correct column names.

File: source/analysis/spam-filter/index.html.md.erb [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+weight: 48
+last_reviewed_on: 2022-07-26
+review_in: 6 months
+---
+
+# Use the GOV.UK feedback spam classifier
+
+The [GOV.UK feedback spam classifier](https://github.com/alphagov/govuk-feedback-spam-classifier) is a [binary classifier](https://machinelearningmastery.com/types-of-classification-in-machine-learning/#:~:text=each%20in%20turn.-,Binary%20Classification,prediction%20(churn%20or%20not)) machine learning model that classifies feedback data as spam or not spam.
+
+We based this model on the [SKLearn architecture](https://scikit-learn.org/stable/).
```
Comment (FelixReillyGDS) [inline source/analysis/spam-filter/index.html.md.erb:12]: Apologies for providing this description, I think I could have described it a bit better. 

I think it's best to say something like "For this project, we fine-tuned a Random Forest Classifier Model downloaded from sklearn.ensemble: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html"

File: source/analysis/spam-filter/index.html.md.erb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    You can use [`venv`](https://docs.python.org/3/tutorial/venv.html#creating-virtual-environments), which is built into Python, or an external tool such as [`pyenv`](https://github.com/pyenv/pyenv).
+
+1. Specify a version of Python to use in the virtual environment. You must use Python 3.6.1 or later.
+
+    How you do this depends on the virtual environment tool you are using.
+
+1. Make sure your Python virtual environment has `pip` installed. If your virtual environment does not have `pip` installed, then [install `pip` manually](https://pip.pypa.io/en/stable/installation/).
+
+1. Activate your virtual environment and use `pip` to install all the Python packages that the spam classifier needs to run:
```
Comment (FelixReillyGDS) [inline source/analysis/spam-filter/index.html.md.erb:30]: I think we need to add a step about the environment variables as this isn't covered by requirements.txt, but @hwrightson @rory-hurley-gds let me know if I'm incorrect.

 In the README we have the instruction to "[load environment variables](https://github.com/alphagov/govuk-feedback-spam-classifier/blob/master/docs/user_guide/loading_environment_variables.md) from .envrc"

**Review by FelixReillyGDS — COMMENTED** [review]
Solid draft of documentation with great instructions for running the model.

My initial comments cover some clarification to do before making it live, and make sure every step is covered for people completely new to the project.

File: source/analysis/spam-filter/index.html.md.erb (hunk 4 above)
Comment (jonathanglassman) [reply source/analysis/spam-filter/index.html.md.erb:30]: I thought that the entire section about secrets and environment variables was not needed, but happy to add something in if I'm wrong

File: source/analysis/spam-filter/index.html.md.erb (hunk 2 above)
Comment (jonathanglassman) [reply source/analysis/spam-filter/index.html.md.erb:45]: good spot

File: source/analysis/spam-filter/index.html.md.erb (hunk 1 above)
Comment (jonathanglassman) [reply source/analysis/spam-filter/index.html.md.erb:63]: just to clarify, the second bullet would say:

"- creates:
    - visualisations to demonstrate the accuracy of the model's predictions in comparison to the manual labels from the training dataset
    - the features that are most 'important' to the model's classification decisions"
?
I.e. it's creating both of these things?

Comment (hwrightson) [reply source/analysis/spam-filter/index.html.md.erb:63]: I would leave it as _"help you assess if the model is classifying spam correctly"_ and add _"and calculates the features that are most 'important' to the model's classification decisions"_

File: source/analysis/spam-filter/index.html.md.erb [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+### Run script 002
+
+Run the [`002_train.py` Python script](https://github.com/alphagov/govuk-feedback-spam-classifier/blob/master/src/002_train.py) in your virtual environment.
+
+This script:
+
+- [trains the spam classifier model](https://en.wikipedia.org/wiki/Training,_validation,_and_test_data_sets) using the specified data set
+- creates visualisations to help you assess if the model is classifying spam correctly
+- sends the visualisations to the [repo `outputs` folder](https://github.com/alphagov/govuk-feedback-spam-classifier/tree/master/outputs)
```
Comment (hwrightson) [inline source/analysis/spam-filter/index.html.md.erb:64]: Can you preemptively update this to link to master/outputs/visualisations, this is a change i'll make to the folder formats in a second. (and update to [repo 'outputs/visualisations' folder]

File: source/analysis/spam-filter/index.html.md.erb [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+One of the visualisations is a [confusion matrix](https://en.wikipedia.org/wiki/Confusion_matrix).
```
Comment (hwrightson) [inline source/analysis/spam-filter/index.html.md.erb:66]: Is it worth saying that the other 3 are representations of different calculations of feature importance?

File: source/analysis/spam-filter/index.html.md.erb [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+### Run script 004
+
+Run the [`004_predict_new_data.py` Python script](https://github.com/alphagov/govuk-feedback-spam-classifier/blob/master/src/004_predict_new_data.py) in your virtual environment.
+
+This script:
+
+- tests the model on the entire data set
+- attaches a label of `spam` or `not spam` and a confidence probability to each sentence
+- outputs the data set with the labels and probabilities to the [repo `data` folder](https://github.com/alphagov/govuk-feedback-spam-classifier/tree/master/data)
```
Comment (hwrightson) [inline source/analysis/spam-filter/index.html.md.erb:86]: Here can we change link to /master/outputs/data and [repo 'outputs/data' folder]. Again, i'll make this change to the folder structure in a second

File: source/analysis/spam-filter/index.html.md.erb (hunk 5 above)
Comment (jonathanglassman) [reply source/analysis/spam-filter/index.html.md.erb:64]: done

File: source/analysis/spam-filter/index.html.md.erb (hunk 6 above)
Comment (jonathanglassman) [reply source/analysis/spam-filter/index.html.md.erb:66]: yes agree

File: source/analysis/spam-filter/index.html.md.erb (hunk 7 above)
Comment (jonathanglassman) [reply source/analysis/spam-filter/index.html.md.erb:86]: done

File: source/analysis/spam-filter/index.html.md.erb (hunk 3 above)
Comment (jonathanglassman) [reply source/analysis/spam-filter/index.html.md.erb:12]: done

File: source/analysis/spam-filter/index.html.md.erb [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+weight: 48
+last_reviewed_on: 2022-07-26
+review_in: 6 months
+---
+
+# Use the GOV.UK feedback spam classifier
+
+The [GOV.UK feedback spam classifier](https://github.com/alphagov/govuk-feedback-spam-classifier) is a [binary classifier](https://machinelearningmastery.com/types-of-classification-in-machine-learning/#:~:text=each%20in%20turn.-,Binary%20Classification,prediction%20(churn%20or%20not)) machine learning model that classifies feedback data as spam or not spam.
+
+To develop this model, we refined a [Random Forest Classifier Model that we downloaded from `sklearn.ensemble`](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html).
```
Comment (rory-hurley-gds) [inline source/analysis/spam-filter/index.html.md.erb:12]: Can we have "using the sklearn.ensemble API"

instead of "that we downloaded from `sklearn.ensemble`"?

File: source/analysis/spam-filter/index.html.md.erb [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+You can use the following documentation to try the feedback spam classifier on your local machine.
+
+## Set up your local machine
+
+Before you can try the feedback spam classifier on your local machine, you must set up your local machine.
+
+1. Set up a [Python virtual environment](https://docs.python.org/3/tutorial/venv.html) on your local machine.
```
Comment (rory-hurley-gds) [inline source/analysis/spam-filter/index.html.md.erb:20]: All the numbering is `1. ...`. Assuming there is a reason for this, but just flagging now.

File: source/analysis/spam-filter/index.html.md.erb [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+If you want to try the feedback spam classifier with a different data set, you must:
+
+- specify where this different data set is in the [params.yaml file in the feedback spam classifier repo](https://github.com/alphagov/govuk-feedback-spam-classifier/blob/master/params.yaml)
+- make sure that this different data set is similar in format and structure to the default data set
+- make any necessary changes to the Python scripts to reflect the different data set's format and structure
+
+### Run script 001
+
+Run the [`001_feature_engineering.py` Python script](https://github.com/alphagov/govuk-feedback-spam-classifier/blob/master/src/001_feature_engineering.py) in your virtual environment.
```
Comment (rory-hurley-gds) [inline source/analysis/spam-filter/index.html.md.erb:117]: Can we add the CLI command here for good measure? See below.

```
python src/001_feature_engineering.py --config params.yaml
```

File: source/analysis/spam-filter/index.html.md.erb [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+This script:
+
+- turns the raw data set into a [Pandas dataframe](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) and identifies relevant columns in the data set
+- searches the text in the Pandas dataframe for features that may indicate that the feedback data is spam (for example, if the feedback text has SQL)
+
+### Run script 002
+
+Run the [`002_train.py` Python script](https://github.com/alphagov/govuk-feedback-spam-classifier/blob/master/src/002_train.py) in your virtual environment.
```
Comment (rory-hurley-gds) [inline source/analysis/spam-filter/index.html.md.erb:126]: Ditto as above

```
python src/002_train.py --config params.yaml
```

File: source/analysis/spam-filter/index.html.md.erb [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+This script:
+
+- [trains the spam classifier model](https://en.wikipedia.org/wiki/Training,_validation,_and_test_data_sets) using the specified data set
+- creates visualisations to help you assess if the model is classifying spam correctly
```
Comment (rory-hurley-gds) [inline source/analysis/spam-filter/index.html.md.erb:131]: "to help users assess model performance" 

instead of 
"if the model is classifying spam correctly"

File: source/analysis/spam-filter/index.html.md.erb [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+- calculates the features that are most important to the model's classification decisions
+- sends the visualisations to the [repo's `outputs/visualisations` folder](https://github.com/alphagov/govuk-feedback-spam-classifier/tree/master/outputs/visualisations)
+
+The visualisations are a [confusion matrix](https://en.wikipedia.org/wiki/Confusion_matrix) and 3 representations of different calculations of [feature importance](https://machinelearningmastery.com/calculate-feature-importance-with-python/).
+
+### Run script 003
+
+Run the [`003_predict.py` Python script](https://github.com/alphagov/govuk-feedback-spam-classifier/blob/master/src/003_predict.py) in your virtual environment.
```
Comment (rory-hurley-gds) [inline source/analysis/spam-filter/index.html.md.erb:139]: Ditto as above

python src/003_predict.py --config params.yaml

File: source/analysis/spam-filter/index.html.md.erb [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+
+This script:
+
+- tests the model on 3 sentences specified in the `003_predict.py` script
+- attaches a label of `spam` or `not spam` and a confidence probability to each sentence
+- outputs the labels and probabilities to the command line
+
+### Run script 004
+
+Run the [`004_predict_new_data.py` Python script](https://github.com/alphagov/govuk-feedback-spam-classifier/blob/master/src/004_predict_new_data.py) in your virtual environment.
```
Comment (rory-hurley-gds) [inline source/analysis/spam-filter/index.html.md.erb:149]: Ditto as above

python src/004_predict_new_data.py --config params.yaml

**Review by rory-hurley-gds — CHANGES_REQUESTED** [review]
Looks great 🌟 

Just a couple of minor requests and a query, but otherwise good to go.

File: source/analysis/spam-filter/index.html.md.erb (hunk 8 above)
Comment (jonathanglassman) [reply source/analysis/spam-filter/index.html.md.erb:12]: agree

File: source/analysis/spam-filter/index.html.md.erb (hunk 9 above)
Comment (jonathanglassman) [reply source/analysis/spam-filter/index.html.md.erb:20]: The tech doc template automatically applies the correct numbering order from the markdown file. Using all `1.`s means that you don't have to update your numbering when you update an ordered list.

File: source/analysis/spam-filter/index.html.md.erb (hunk 10 above)
Comment (jonathanglassman) [reply source/analysis/spam-filter/index.html.md.erb:117]: agree

File: source/analysis/spam-filter/index.html.md.erb (hunk 11 above)
Comment (jonathanglassman) [reply source/analysis/spam-filter/index.html.md.erb:126]: agree

File: source/analysis/spam-filter/index.html.md.erb (hunk 12 above)
Comment (jonathanglassman) [reply source/analysis/spam-filter/index.html.md.erb:131]: agree

File: source/analysis/spam-filter/index.html.md.erb (hunk 9 above)
Comment (rory-hurley-gds) [reply source/analysis/spam-filter/index.html.md.erb:20]: Good to know - thanks for clarifying

**Review by rory-hurley-gds — APPROVED** [review]
Ready to merge
